File: app/src/main/java/nl/waywayway/ahn/LocationProvider.java
package nl.waywayway.ahn;

import android.content.*;
import android.location.*;
import android.os.*;
import android.util.*;
import com.google.android.gms.common.*;
import com.google.android.gms.common.api.*;
import com.google.android.gms.location.*;
import com.google.android.gms.maps.model.*;

import com.google.android.gms.location.LocationListener;

public class LocationProvider
implements 
LocationListener,
GoogleApiClient.ConnectionCallbacks,
GoogleApiClient.OnConnectionFailedListener
{
	private GoogleApiClient googleApiClient;
	private float zoom;
	private LocationProvider thisInstance;
	private Context context;
	Handler handler;

	public LocationProvider(Context context)
	{
		this.context = context;
		this.googleApiClient = createGoogleApiClient();
		this.thisInstance = this;
	}

	private GoogleApiClient createGoogleApiClient()
	{
		googleApiClient = new GoogleApiClient
			.Builder(context)
			.addConnectionCallbacks(this)
			.addOnConnectionFailedListener(this)
			.addApi(LocationServices.API)
			.build();
			
		return googleApiClient;
	}
	
	// Zoom in op huidige locatie of midden van Nederland bij eerste opstart app
	// timeOut in seconden
	public void zoomToCurrentOrStandardLocation(final LatLng standardLocation, int timeOut, final float zoom)
	{
		//Log.i("HermLog", "zoomToCurrentOrStandardLocation");
		Location currentLocation = getCurrentLocation();

		if (currentLocation != null)
		{
			handleLocation(currentLocation, zoom);
		}
		else
		{
			// wacht x seconden op locatie
			LocationRequest locationRequest = LocationRequest.create()
				.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
				.setNumUpdates(1)
				.setExpirationDuration(1000 * timeOut);

			LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, this);

			Runnable expiredRunnable = new Runnable()
			{
				@Override
				public void run()
				{
					//Log.i("HermLog", "expiredRunnable");
					LocationServices.FusedLocationApi.removeLocationUpdates(googleApiClient, thisInstance);
					locationUnavailable(standardLocation, zoom);
				}
			};
			
			handler = new Handler();
			handler.postDelayed(expiredRunnable, 1000 * timeOut);
		}

		//Toast.makeText(this, getResources().getString(R.string.device_location_not_available_message), Toast.LENGTH_SHORT).show();
	}

	@Override
	public void onLocationChanged(Location location)
	{
		//Log.i("HermLog", "onLocationChanged");
		LocationServices.FusedLocationApi.removeLocationUpdates(googleApiClient, this);
		handler.removeCallbacksAndMessages(null);
        handleLocation(location, zoom);
	}

	public Location getCurrentLocation()
	{
		//Log.i("HermLog", "getCurrentLocation");
		// Vraag huidige locatie op
		Location lastLocation = LocationServices.FusedLocationApi
			.getLastLocation(googleApiClient);

		return lastLocation;
	}
	
	public void connect()
	{
		googleApiClient.connect();
	}
	
	public void disconnect()
	{
		googleApiClient.disconnect();
	}
	
	// Methods bedoeld om te overschrijven bij intantiatie
	@Override
	public void onConnected(Bundle bundle)
	{}

	@Override
	public void onConnectionSuspended(int code)
	{}
	
	@Override
	public void onConnectionFailed(ConnectionResult connectionResult)
	{}
	
	public void handleLocation(Location location, float zoom)
	{}

	public void locationUnavailable(LatLng standardLocation, float zoom)
	{}
}

File: app/src/main/java/nl/waywayway/ahn/LocationProvider.kt
package nl.waywayway.ahn

import android.content.Context
import android.content.IntentSender
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.widget.Toast

import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.*
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.tasks.Task

// Class voor opvragen van en zoomen naar locatie toestel of standaardlocatie

class LocationProvider(val context: Context, val gMap: GoogleMap, var standardLocation: LatLng) : LocationCallback() {
    var fusedLocationClient: FusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
    val standardZoomLevel: Float = context.resources.getInteger(R.integer.standard_zoom_level).toFloat()
    val timeOut: Long = 1000
    val numberOfUpdates: Int = 1
    val priorityCode: Int = LocationRequest.PRIORITY_HIGH_ACCURACY
    var zoomToStandardLocationAsDefault: Boolean = false

    fun zoomToDeviceOrStandardLocation() {
        // Zoom in op huidige locatie of midden van Nederland bij eerste opstart app
        zoomToStandardLocationAsDefault = true
        zoomToDeviceLocation()
    }

    fun zoomToDeviceLocation() {
        requestLocationUpdate()
        /*
        fusedLocationClient.lastLocation
                .addOnSuccessListener { location: Location? ->
                    if (location != null) {
                        // Locatie beschikbaar
                        Log.i("HermLog", "zoomToDeviceLocation(): laatste locatie beschikbaar");
                        val lat = location.getLatitude()
                        val lon = location.getLongitude()
                        zoomToLocation(LatLng(lat, lon), standardZoomLevel)
                    } else {
                        // Locatie niet beschikbaar, vraag locatie op
                        Log.i("HermLog", "zoomToDeviceLocation(): laatste locatie niet beschikbaar");
                        requestLocationUpdate()
                    }
                }*/
    }

    fun requestLocationUpdate() {
        val locationRequest = LocationRequest.create()?.apply {
            setExpirationDuration(timeOut)
            numUpdates = numberOfUpdates
            priority = priorityCode
        }

        fusedLocationClient.requestLocationUpdates(locationRequest, this, null)
    }

    override fun onLocationResult(locationResult: LocationResult?) {
        fusedLocationClient.removeLocationUpdates(this)

        if (locationResult == null) {
            // Locatie definitief niet beschikbaar
            Log.i("HermLog", "onLocationResult: locatie update opgevraagd maar niet beschikbaar")
            Toast.makeText(context, context.resources.getString(R.string.device_location_not_available_message), Toast.LENGTH_SHORT).show()
            if (zoomToStandardLocationAsDefault) zoomToLocation(standardLocation, standardZoomLevel)
            zoomToStandardLocationAsDefault = false
        } else {
            // Locatie opgevraagd en beschikbaar
            Log.i("HermLog", "onLocationResult: locatie update opgevraagd en beschikbaar")
            val location: Location? = locationResult.locations[0]
            val lat = location?.latitude
            val lon = location?.longitude
            Log.i("HermLog", "onLocationResult: lat: " + lat)
            Log.i("HermLog", "onLocationResult: long: " + lon)
            zoomToLocation(LatLng(lat!!, lon!!), standardZoomLevel)
        }
    }

    fun zoomToLocation(location: LatLng, zoomLevel: Float) {
        gMap.animateCamera(CameraUpdateFactory.newLatLngZoom(location, zoomLevel))
    }
}